fedmsg/commands/check.py

```diff
@@ -0,0 +1,101 @@
+import zmq
+
+from fedmsg.config import load_config
+
+
+TIMEOUT_HELP = _("The number of seconds to wait for a heartbeat from the fedmsg hub."
+                 " The default is 10 seconds.")
+CONSUMER_HELP = _("The name of the consumer to check the status of.")
+PRODUCER_HELP = _("The name of the producer to check the status of.")
+
+
+@click.command()
+@click.option('--timeout', default=10, type=click.INT, help=TIMEOUT_HELP)
+@click.option('--consumer', '-c', multiple=True, help=CONSUMER_HELP)
+@click.option('--producer', '-p', multiple=True, help=PRODUCER_HELP)
+def check(timeout, consumer=None, producer=None):
+    """This command is used to check the status of consumers and producers.
+
+    If no consumers and producers are provided, the status of all consumers and producers
+    is printed.
+
+    Args:
+        timeout (int): The length of time in seconds to wait for a ZMQ message
+            on the monitor socket.
+        consumer (tuple): A list of consumer names to check the status of.
+        producer (tuple): A list of producer names to check the status of.
+    """
+    # It's weird to say --consumers, but there are multiple, so rename the variables
+    consumers, producers = consumer, producer
+
+    config = load_config()
+    endpoint = config.get('moksha.monitoring.socket')
+    if not endpoint:
+        raise click.ClickException('No monitoring endpoint has been configured: '
+                                   'please set "moksha.monitoring.socket"')
+
+    context = zmq.Context.instance()
+    socket = context.socket(zmq.SUB)
+    # ZMQ takes the timeout in milliseconds
+    socket.set(zmq.RCVTIMEO, timeout * 1000)
+    socket.subscribe(b'')
+    socket.connect(endpoint)
+
+    try:
+        message = socket.recv_json()
+    except zmq.error.Again:
+        raise click.ClickException(
+            'Failed to receive message from the monitoring endpoint ({}) in {} '
+            'seconds.'.format(endpoint, timeout))
+
+    if not consumers and not producers:
+        click.echo('No consumers or producers specified so all will be shown.')
+    else:
+        missing = False
+        uninitialized = False
+        for messager_type, messagers in (('consumers', consumers), ('producers', producers)):
+            active = {c['name']: c for c in message[messager_type]}
+            for messager in messagers:
+                if messager not in active:
+                    click.echo('"{}" is not active!'.format(messager), err=True)
+                    missing = True
+                else:
+                    if active[messager]['initialized'] is not True:
+                        click.echo('"{}" is not initialized!'.format(messager), err=True)
+                        uninitialized = True
+
+        if missing:
+            raise click.ClickException('Some consumers and/or producers are missing!')
+        elif uninitialized:
+            raise click.ClickException('Some consumers and/or producers are uninitialized!')
```

Comment:
Might you want this to be an ```if``` instead of an ```elif```? I.e., if some are missing and others are uninitialized, it'd be nice to report both to the user instead of one or the other. Of course, that breaks the final ```else```, so you probably need to wrap the above into an overall ```if missing or unitialized``` and then do inner ```if```s…